# Tarea 1: Introducción, Modelos de Espacio Vectorial, Recuperación de Información y Modelos de Lenguaje

**Procesamiento de Lenguaje Natural (CC6205-1 - Otoño 2025)**

## Tarjeta de Identificación

- **Nombre(s):** ```Israel Astudillo, Fabrizzio Pezolla, Rodrigo Molina```
- **Fecha límite de entrega 📆:** 15 de abril de 2025
- **Tiempo estimado de dedicación:** 4 horas

---

## 📌 Instrucciones

¡Bienvenid@s a la primera tarea del curso de *Natural Language Processing* (NLP)!

El objetivo de esta tarea es evaluar los conceptos teóricos de las primeras semanas de clases, centrándose en:

- **Recuperación de Información (IR)**
- **Modelos de Espacio Vectorial**
- **Modelos de Lenguaje**

Si aún no has revisado el contenido correspondiente, se recomienda consultar las referencias disponibles al final del documento.

### 📢 Consideraciones generales

✅ La tarea debe realizarse en **grupos de hasta 3 personas**.

✅ La entrega debe realizarse a través de **U-Cursos**, a más tardar en la fecha estipulada. **No se aceptarán entregas atrasadas.**

✅ El formato de entrega es este mismo **Jupyter Notebook**.

✅ Su código será ejecutado al momento de la revisión. **Verifiquen que no tenga errores de compilación.**

✅ Es obligatorio completar la **Tarjeta de Identificación**. **No se asignará nota sin ella.**

---

## 📚 Material de Referencia

### 📄 Diapositivas del curso

- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Modelos de Espacio Vectorial y Recuperación de Información](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)
- [Modelos Probabilísticos de Lenguaje](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-PLM.pdf)

### 📺 Videos del curso

- **Introducción**: [Parte 1](https://www.youtube.com/watch?v=HEKTNOttGvU) | [Parte 2](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- **Recuperación de Información**: [Parte 1](https://www.youtube.com/watch?v=FXIVClF370w&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3) | [Parte 2](https://www.youtube.com/watch?v=f8nG1EMmPZk&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3)
- **Modelos Probabilísticos de Lenguaje**: [Parte 1](https://www.youtube.com/watch?v=9E2jJ6kcb4Y&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3) | [Parte 2](https://www.youtube.com/watch?v=ZWqbEQXLra0&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=5) | [Parte 3](https://www.youtube.com/watch?v=tsumFqwFlaA&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=6) | [Parte 4](https://www.youtube.com/watch?v=s3TWdv4sqkg&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJq)

---

📌 **Recuerda:** La claridad y organización en la entrega son clave para una mejor evaluación. Incluir analisis apropiado. ¡Mucho éxito! 🚀

## P1. Tokenización

En el primer ejercicio veremos la dificultad de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo.

In [1]:
# En caso de desarrollar la tarea desde colab, con el siguiente código podemos cargar los archivos desde drive:

try:
    from google.colab import drive

    drive.mount("/content/drive", force_remount=True)
    path = '/content/drive/MyDrive/nlp/oh_algoritmo.txt'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook


In [2]:
try:
    # Abre el archivo en modo lectura ("r")
    with open("oh_algoritmo.txt", "r") as archivo:
        # Lee el contenido del archivo
        texto = archivo.read()
        # Imprime el contenido
        print(texto)
except FileNotFoundError:
    print("El archivo no se encuentra.")
except Exception as e:
    print("Ocurrió un error:", e)

[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]

[Refrán: Jorge Drexler]
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?

[Estribillo: Jorge Drexler]
Dime qué debo cantar
Oh, algoritmo
Sé que lo sabes mejor
Incluso que yo mismo

[Verso 1: Nora Erez]
Wait, what's that money that you spent?
What's that sitting on your plate?
Do you want what you've been fed?
Are you the fish or bait?
Mmm, I'm on the top of the roof and I feel like a jail
Rather not pay the bail
To dangerous people with blood on their faces
So I'm sharing a cell with the masses
The underground always strive for the main
Streaming like Grande's big-ass ring
Screaming: I'll write you out my will
Conscious is free, but not the will
Conscious is free, but not the will
You

Fuente: https://genius.com/Jorge-drexler-oh-algoritmo-lyrics

### Pregunta 1.a (0.25 puntos)

Diseñe una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Es libre de elegir la forma de tokenizar mientras no utilice librerías con tokenizadores ya implementados. Puede utilizar la librería **re** importada para trabajar símbolos. Explique su razonamiento.


In [3]:
import re

In [4]:
def get_tokens(texto):
    ### Aquí inicia tu código ###
    """
    get_tokens recibe un texto y entrega una lista con sus tokens.
    [str]->list
    """
    assert type(texto) == str, "El texto debe ser un string"
    tokens = re.findall(r'\w+|[^\w\s]', texto)
    return tokens
    ### Aquí termina tu código ##

In [5]:
# Función get_tokens alternativa
def get_tokens_alt(texto):
    ### Aquí inicia tu código ###
    """
    get_tokens recibe un texto y entrega una lista con sus tokens.
    [str]->list
    """
    assert type(texto) == str, "El texto debe ser un string"
    tokens = texto.split(sep=" ")
    for i in range(len(tokens)):
      token = tokens[i]
      token_limpio = re.sub(r'[^a-zA-Z]', '', token).lower()
      tokens[i] = token_limpio
    return tokens

In [6]:
tokens = get_tokens_alt(texto)
tokens

['letra',
 'de',
 'oh',
 'algoritmo',
 'ft',
 'nora',
 'erezrefrn',
 'jorge',
 'drexlerquin',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quieroquin',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quieroquin',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quieroquin',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quieroquin',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quieroquin',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quieroestribillo',
 'jorge',
 'drexlerdime',
 'qu',
 'debo',
 'cantaroh',
 'algoritmos',
 'que',
 'lo',
 'sabes',
 'mejorincluso',
 'que',
 'yo',
 'mismoverso',
 '',
 'nora',
 'erezwait',
 'whats',
 'that',
 'money',
 'that',
 'you',
 'spentwhats',
 'that',
 'sitting',
 'on',
 'your',
 'platedo',
 'you',
 'want',
 'what',
 'youve',
 'been',
 'fedare',
 'you',
 'the',
 'fish',
 'or',
 'baitmmm',
 'im',
 'on',
 't

### Pregunta 1.b (0.25 puntos)
Explique su implementación aquí:
> La implementación utiliza la función .findall de la librería re, que encuentra todas las apariciones del patrón expresado en el primer argumento, en forma de expresión regular, en el texto insertado en el segundo argumento. La expresión regular utilizada es `r'\w+|[^\w\s]'`. La letra r sirve para que la expresión no sea interpretada por python directamente como un comando y no utilice los símbolos de escape como "\" directamente, sino que la expresión sea entregada tal cual. La expresión se divide en dos partes por el símbolo "|" que significa or lógico. La primera parte `\w+` significa uno o más de un caracter alfanumérico de cualquier tipo, incluyendo guiones bajos. Esto captura conjuntos de letras como palabras y los distingue de cualquier otro separador como espacios, puntuaciones, entre otros. La segunda parte `[^\w\s]'` define una clase de caracteres con "[ ]", siendo la clase definida la negación (denotado con "^") de los caracteres alfanuméricos y espacios blancos de cualquier tipo como tab, saltos de línea, entre otros. Es decir, esta clase captura todo lo que no sean caracteres alfanuméricos, guiones bajos y espacios blancos, lo cual deja solo los símbolos de puntuación en general distinguidos como caracteres separados independientes. Así, en la salida de la función esperamos palabras y símbolos de puntuación, que es el resultado que se obtuvo.

> La implementación utiliza la función .split nativa de python para separar el string en strings más pequeños utilizando únicamente los espacios como separadores, retornando el resultado. Luego elimina todas los caracteres que no sean letras y transforma a minúscula el texto.

Implementación con la libreria NLTK

In [7]:
from nltk.tokenize import wordpunct_tokenize
nltk_tokens = wordpunct_tokenize(texto)
nltk_tokens

['[',
 'Letra',
 'de',
 '"¡',
 'Oh',
 ',',
 'Algoritmo',
 '!"',
 'ft',
 '.',
 'Nora',
 'Erez',
 ']',
 '[',
 'Refrán',
 ':',
 'Jorge',
 'Drexler',
 ']',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '[',
 'Estribillo',
 ':',
 'Jorge',
 'Drexler',
 ']',
 'Dime',
 'qué',
 'debo',
 'cantar',
 'Oh',
 ',',
 'algoritmo',
 'Sé',
 'que',
 'lo',
 'sabes',
 'mejor',
 'Incluso',
 'que',
 'yo',
 'mismo',
 '[',
 'Verso',
 '1',
 ':',
 'Nora',
 

### Pregunta 1.c (0.5 puntos)
¿Qué diferencias y similitudes encontrase al comparar la función de tokenización creada manualmente por ti contra la implementación de NLTK, al tokenizar la letra de la canción "Oh, algoritmo"?

> La diferencia con la implementación de la librería es que no se reconocen caracteres especiales y algunas palabras no se diferencian correctamente. Por ejemplo el token "drexlerpreestribillo" no es un token adecuado en comparación al resultado de la librería, por lo que no es un muy buen resultado.

## P2. Stemming y Stopwords

En esta sección debera implementar funciones de stemming y stopwords basado en lo visto en clase. En la siguiente celda tiene el corpus que usara en esta sección:

In [8]:
# Corpus en español
corpus_espanol = [
    "¿Quién quiere que yo quiera lo que creo que quiero?",
    "Dime qué debo cantar",
    "Sé que lo sabes mejor"
]

# Corpus en inglés
corpus_ingles = [
    "What's that sitting on your plate?",
    "Do you want what you've been fed?",
    "Are you the fish or bait?"
]

### Pregunta 2.a (0.5 puntos)
Implemente una función **`get_vocab()`** que extraiga los tokens de un corpus. Puede utilizar la función de la sección anterior.

In [9]:
def get_vocab(corpus):
  ### Aquí inicia tu código ###
  """
  get_tokens recibe un texto y entrega una lista con sus tokens.
  [str]->list
  """
  assert type(corpus) == str, "El texto debe ser un string"
  tokens = re.findall(r'\w+|[^\w\s]', corpus)
  return tokens
  ### Aquí termina tu código ###

In [10]:
vocab_espanol = [get_vocab(vocab) for vocab in corpus_espanol]
vocab_espanol = [palabra for sublista in vocab_espanol for palabra in sublista]
vocab_espanol

['¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 'Dime',
 'qué',
 'debo',
 'cantar',
 'Sé',
 'que',
 'lo',
 'sabes',
 'mejor']

Resultado esperado (el orden puede variar):
```
['yo', 'debo', 'creo', 'Dime', 'lo', 'cantar', 'mejor', 'Sé', 'que', 'quiere', 'quiero', 'sabes', 'Quién', 'quiera', 'qué']
```

In [11]:
vocab_ingles = [get_vocab(vocab) for vocab in corpus_ingles]
vocab_ingles = [palabra for sublista in vocab_ingles for palabra in sublista]
vocab_ingles

['What',
 "'",
 's',
 'that',
 'sitting',
 'on',
 'your',
 'plate',
 '?',
 'Do',
 'you',
 'want',
 'what',
 'you',
 "'",
 've',
 'been',
 'fed',
 '?',
 'Are',
 'you',
 'the',
 'fish',
 'or',
 'bait',
 '?']

Resultado esperado:
```
['fed', 'been', 'or', 'want', 'plate', 'the', 've', 'your', 's', 'you', 'what', 'Are', 'bait', 'What', 'fish', 'that', 'sitting', 'Do', 'on']
```

### Pregunta 2.b (0.5 puntos)
Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario preprocesado. Explique las reglas de stemming y elección de stopwords:

    Explique sus reglas aquí: Primero se aplicará stemming para homogenizar las palabras en el corpus, de forma de luego contar la frecuencia de cada palabra para identificar stopwords sobre los tokens homogenizados.
    
    Para hacer stemming en el corpus en español identificamos los verbos conjugados y les quitamos sus extremos del final. Podemos lograr esto eliminando las vocales o vocales + "s" del final de palabras largas o de más de 4 caracteres.

    En el corpus en inglés podemos eliminar las terminaciones "ing" en este corpus en específico. Podríamos eliminar también las "s" al final de las palabras.


In [12]:
def pre_processing(vocabulario, idioma):
    ### Aquí inicia tu código ###
    vowels = ["a", "e", "i", "o", "u"]
    new_vocab = vocabulario.copy()

    # Eliminar tokens no alfanuméricos
    new_vocab = [p for p in new_vocab if p.isalnum()]

    if idioma == "espanol":
        # Stemming español
        for i in range(len(new_vocab)):
            palabra = new_vocab[i]
            if len(palabra) >= 4:
                if palabra[-1] == "s" and palabra[-2] in vowels:
                    palabra = palabra[:-2]  # vocal + s
                elif palabra[-1] in vowels:
                    palabra = palabra[:-1]
                elif palabra[-1] == "s":
                    palabra = palabra[:-1]
            new_vocab[i] = palabra

    elif idioma == "ingles":
        # Stemming inglés
        for i in range(len(new_vocab)):
            palabra = new_vocab[i]
            if palabra.endswith("ing") and len(palabra) > 4:
                palabra = palabra[:-3]
            elif palabra.endswith("s") and len(palabra) > 3:
                palabra = palabra[:-1]
            new_vocab[i] = palabra

    else:
        raise Exception("Idioma no soportado")

    # Contar frecuencias para eliminar Stopwords
    frec = {}
    for palabra in new_vocab:
        frec[palabra] = frec.get(palabra, 0) + 1

    # Eliminar stopwords: palabras con >1 ocurrencia y largo <= 3
    new_vocab = [p for p in new_vocab if not (frec[p] > 1 and len(p) <= 3)]
    return new_vocab
    ### Aquí termina tu código ##

In [13]:
# Aplicar preprocesamiento a los vocabularios de ejemplo con NLTK
vocab_procesado_espanol = pre_processing(vocab_espanol, 'espanol')
vocab_procesado_ingles = pre_processing(vocab_ingles, 'ingles')

# Mostrar resultados
print("Vocabulario procesado en español:", vocab_procesado_espanol, "\n")
print("Vocabulario procesado en inglés:", vocab_procesado_ingles, "\n")

Vocabulario procesado en español: ['Quién', 'quier', 'yo', 'quier', 'cre', 'quier', 'Dim', 'qué', 'deb', 'cantar', 'Sé', 'sab', 'mejor'] 

Vocabulario procesado en inglés: ['What', 's', 'that', 'sitt', 'on', 'your', 'plate', 'Do', 'want', 'what', 've', 'been', 'fed', 'Are', 'the', 'fish', 'or', 'bait'] 



## P3. Bag of Words (0.5 puntos)
Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:


In [14]:
d0 = 'El pájaro come semillas'
d1 = 'El pájaro se despierta y canta'
d2 = 'El pájaro canta y come semillas'
d3 = 'El pez come y nada en el agua'
d4 = 'El pez empieza a nadar'
d5 = 'El pez come alimento'
corpus = [d0, d1, d2, d3, d4, d5]

El objetivo da las siguientes secciones es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica **TF-IDF**.

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es la de **Bag of Words**, método mediante el cual se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que recibe como input un arreglo de documentos y devuelve un dataframe de pandas con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representan las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el BoW de un documento.

***Disclaimer: el orden de los resultados pueden variar.***


Por ejemplo para el siguiente corpus:

```
corpus = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |




In [15]:
import pandas as pd

Implementar función `bag_of_words()`

In [16]:
def bag_of_words(corpus):
    ### Aquí inicia tu código ###
    from collections import Counter

    bow_repr = {
        f'd{idx}': Counter(doc.split())
        for idx, doc in enumerate(corpus)
        }

    return pd.DataFrame(bow_repr).T.fillna(0).astype(int)

    ### Aquí termina tu código ##

In [17]:
dataset_bow = bag_of_words(corpus)
dataset_bow

El  pájaro  come  semillas  se  despierta  y  canta  pez  nada  en  el  \
d0   1       1     1         1   0          0  0      0    0     0   0   0   
d1   1       1     0         0   1          1  1      1    0     0   0   0   
d2   1       1     1         1   0          0  1      1    0     0   0   0   
d3   1       0     1         0   0          0  1      0    1     1   1   1   
d4   1       0     0         0   0          0  0      0    1     0   0   0   
d5   1       0     1         0   0          0  0      0    1     0   0   0   

    agua  empieza  a  nadar  alimento  
d0     0        0  0      0         0  
d1     0        0  0      0         0  
d2     0        0  0      0         0  
d3     1        0  0      0         0  
d4     0        1  1      1         0  
d5     0        0  0      0         1

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

## P4. TF-IDF

### 4.a TF (0.25 puntos)

Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia $\max_i({\text{tf}_{i,j}})$, donde
$i$ corresponde al índice de las filas (BoW) y $j$ al de las columnas (palabras). Es decir, dividir cada BoW sobre la cantidad de veces de la palabra que aparezca más veces en ese vector.


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{\max_i({\text{tf}_{i,j})}}$$

Implemente la función `calc_tf(dataset_bow)`, que entrega la matriz de TF normalizada del BoW del dataset.

In [18]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    return dataset_bow.div(dataset_bow.max(axis='columns'), axis='index')
    ### Aquí termina tu código ##

In [19]:
tf = calc_tf(dataset_bow)
tf

El  pájaro  come  semillas   se  despierta    y  canta  pez  nada   en  \
d0  1.0     1.0   1.0       1.0  0.0        0.0  0.0    0.0  0.0   0.0  0.0   
d1  1.0     1.0   0.0       0.0  1.0        1.0  1.0    1.0  0.0   0.0  0.0   
d2  1.0     1.0   1.0       1.0  0.0        0.0  1.0    1.0  0.0   0.0  0.0   
d3  1.0     0.0   1.0       0.0  0.0        0.0  1.0    0.0  1.0   1.0  1.0   
d4  1.0     0.0   0.0       0.0  0.0        0.0  0.0    0.0  1.0   0.0  0.0   
d5  1.0     0.0   1.0       0.0  0.0        0.0  0.0    0.0  1.0   0.0  0.0   

     el  agua  empieza    a  nadar  alimento  
d0  0.0   0.0      0.0  0.0    0.0       0.0  
d1  0.0   0.0      0.0  0.0    0.0       0.0  
d2  0.0   0.0      0.0  0.0    0.0       0.0  
d3  1.0   1.0      0.0  0.0    0.0       0.0  
d4  0.0   0.0      1.0  1.0    1.0       0.0  
d5  0.0   0.0      0.0  0.0    0.0       1.0

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

### 4.b IDF (0.5 puntos)

Implementar `calc_idf(dataset_bow)`. Ésta debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el cálculo de cada idf por palabra.

Recordar que $\text{idf}_{t_i} = \log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ número de documentos que contienen la palabra $t_i$

In [20]:
import numpy as np

In [21]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    N = dataset_bow.shape[0]
    doc_counts = (dataset_bow > 0).sum(axis=0)
    idf = np.log10(N / doc_counts)
    return idf.to_dict()
    ### Aquí termina tu código ##

In [22]:
idf = calc_idf(dataset_bow)
idf

{'El': 0.0,
 'pájaro': 0.3010299956639812,
 'come': 0.17609125905568124,
 'semillas': 0.47712125471966244,
 'se': 0.7781512503836436,
 'despierta': 0.7781512503836436,
 'y': 0.3010299956639812,
 'canta': 0.47712125471966244,
 'pez': 0.3010299956639812,
 'nada': 0.7781512503836436,
 'en': 0.7781512503836436,
 'el': 0.7781512503836436,
 'agua': 0.7781512503836436,
 'empieza': 0.7781512503836436,
 'a': 0.7781512503836436,
 'nadar': 0.7781512503836436,
 'alimento': 0.7781512503836436}

Solución esperada:
```
{'El': 0.0,
 'pájaro': 0.3010299956639812,
 'despierta': 0.7781512503836436,
 'el': 0.7781512503836436,
 'come': 0.17609125905568124,
 'a': 0.7781512503836436,
 'nadar': 0.7781512503836436,
 'se': 0.7781512503836436,
 'en': 0.7781512503836436,
 'y': 0.3010299956639812,
 'alimento': 0.7781512503836436,
 'semillas': 0.47712125471966244,
 'pez': 0.3010299956639812,
 'empieza': 0.7781512503836436,
 'canta': 0.47712125471966244,
 'agua': 0.7781512503836436,
 'nada': 0.7781512503836436}
 ```

### 4.c TF-IDF (0.25 puntos)
Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [23]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###
    return tf * idf
    ### Aquí termina tu código ##

In [24]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

El   pájaro      come  semillas        se  despierta        y     canta  \
d0  0.0  0.30103  0.176091  0.477121  0.000000   0.000000  0.00000  0.000000   
d1  0.0  0.30103  0.000000  0.000000  0.778151   0.778151  0.30103  0.477121   
d2  0.0  0.30103  0.176091  0.477121  0.000000   0.000000  0.30103  0.477121   
d3  0.0  0.00000  0.176091  0.000000  0.000000   0.000000  0.30103  0.000000   
d4  0.0  0.00000  0.000000  0.000000  0.000000   0.000000  0.00000  0.000000   
d5  0.0  0.00000  0.176091  0.000000  0.000000   0.000000  0.00000  0.000000   

        pez      nada        en        el      agua   empieza         a  \
d0  0.00000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
d1  0.00000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
d2  0.00000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
d3  0.30103  0.778151  0.778151  0.778151  0.778151  0.000000  0.000000   
d4  0.30103  0.000000  0.000000  0.000000  0.000000  0.778151  0.778151   
d5  0.30103  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   

       nadar  alimento  
d0  0.000000  0.000000  
d1  0.000000  0.000000  
d2  0.000000  0.000000  
d3  0.000000  0.000000  
d4  0.778151  0.000000  
d5  0.000000  0.778151

Solución esperada:

|    |   El |   pájaro |   despierta |       el |     come |        a |    nadar |       se |       en |       y |   alimento |   semillas |     pez |   empieza |    canta |     agua |     nada |
|:---|-----:|---------:|------------:|---------:|---------:|---------:|---------:|---------:|---------:|--------:|-----------:|-----------:|--------:|----------:|---------:|---------:|---------:|
| d0 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0        |   0.477121 | 0       |  0        | 0        | 0        | 0        |
| d1 |    0 |  0.30103 |    0.778151 | 0        | 0        | 0        | 0        | 0.778151 | 0        | 0.30103 |   0        |   0        | 0       |  0        | 0.477121 | 0        | 0        |
| d2 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0.30103 |   0        |   0.477121 | 0       |  0        | 0.477121 | 0        | 0        |
| d3 |    0 |  0       |    0        | 0.778151 | 0.176091 | 0        | 0        | 0        | 0.778151 | 0.30103 |   0        |   0        | 0.30103 |  0        | 0        | 0.778151 | 0.778151 |
| d4 |    0 |  0       |    0        | 0        | 0        | 0.778151 | 0.778151 | 0        | 0        | 0       |   0        |   0        | 0.30103 |  0.778151 | 0        | 0        | 0        |
| d5 |    0 |  0       |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0.778151 |   0        | 0.30103 |  0        | 0        | 0        | 0        |

## P5. Cosine-similarity (0.25 puntos)
Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica.

Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos. Concluya cuáles son los dos documentos más similares.

In [25]:
import numpy as np

def cosine_similarity(v1, v2):
    ### Aquí inicia tu código ###
    return np.dot(v1,v2)/(np.linalg.norm(v1)*np.linalg.norm(v2))
    ### Aquí termina tu código ##

In [26]:
similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity

for i in range(6):
  mask = [k != i for k in range(6)]
  j = np.argmax(similarity_matrix[i][mask])

  print(corpus[i])
  print("> Mas similar:", np.array(corpus)[mask][j])
  print("> Similitud:", similarity_matrix[i][mask][j], "\n")

El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520414 

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.3932010182312894 

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520414 

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406168 

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.07695078406752713 

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.08787900372173231 



Solución esperada:
```
El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520414

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.39320101823128945

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520414

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406168

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.07695078406752713

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.0878790037217323
```

## P6 N-gramas (0.75 punto)

En esta sección debera determinar los n-gramas del la cancion "Oh algoritmo".

### 6.a Corpus de entrenamiento y test (0.25 puntos)

En esta subsección debera definir el conjunto de entrenamiento y test de un corpus. Eliga una particion del 80% y 20% del texto.

In [27]:
try:
    # Abre el archivo en modo lectura ("r")
    with open("oh_algoritmo.txt", "r") as archivo:
        # Lee el contenido del archivo
        texto = archivo.read()
        # Imprime el contenido
        print(texto[:50])
except FileNotFoundError:
    print("El archivo no se encuentra.")
except Exception as e:
    print("Ocurrió un error:", e)

[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]

[Refrá


Defina una funcion `get_sentences()` que entregue todas las oraciones del corpus que contengan al menos una palabra.

In [28]:
def get_sentences(texto):
  ## Implementar aquí
  return [x for x in texto.split('\n') if len(x.split(' '))>0 and len(x.split(' ')[0])>0]
  ##

In [29]:
oraciones_limpias = get_sentences(texto)
oraciones_limpias

['[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]',
 '[Refrán: Jorge Drexler]',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '[Estribillo: Jorge Drexler]',
 'Dime qué debo cantar',
 'Oh, algoritmo',
 'Sé que lo sabes mejor',
 'Incluso que yo mismo',
 '[Verso 1: Nora Erez]',
 "Wait, what's that money that you spent?",
 "What's that sitting on your plate?",
 "Do you want what you've been fed?",
 'Are you the fish or bait?',
 "Mmm, I'm on the top of the roof and I feel like a jail",
 'Rather not pay the bail',
 'To dangerous people with blood on their faces',
 "So I'm sharing a cell with the masses",
 'The underground always strive for the main',
 "Streaming like Grande's big-ass ring",
 "Screaming: I'll wr

Debería obtener en total 87 oraciones.

In [30]:
len(oraciones_limpias) == 87

True

Ahora definiremos el conjunto de entrenamiento y prueba para las oraciones:

In [31]:
split = int(len(oraciones_limpias) * 0.8)
train_corpus = oraciones_limpias[:split]
test_corpus = oraciones_limpias[split:]

len(train_corpus), len(test_corpus)

(69, 18)

### 6.b Estimación de N-gramas (0.5 puntos)

Defina una función que reciba una lista de oraciones de un corpus y un N que indique el tamaño de los N-gramas. La función debe retornar un diccionario de Python donde la llave es un token (o palabra) y el valor es la cantidad de veces que ocurre el token, es decir, la frecuencia. En el caso de N-gramas con N mayor a 1 (como bi-gramas o tri-gramas) debe añadir un token especial al inicio o final de cada oración según corresponda (ver clases del curso).

In [32]:
from nltk.tokenize import word_tokenize

import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\rodri\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [33]:
from collections import Counter

In [34]:
def n_grams(corpus, n=3):
  ## Implementar aquí

  assert n in [1,2,3]

  token_lists = [
      word_tokenize(f"BOS {sent.strip()} EOS") if n>1
      else word_tokenize(sent)
      for sent in corpus
  ]

  all_ngrams = [
      tuple(tokens[i : i+n])
      for tokens in token_lists
      for i in range(len(tokens) - n + 1)
  ]

  freqs = Counter(all_ngrams) if n>1 else Counter(all_ngrams + [])

  return dict(freqs)

  ##

In [35]:
n_grams(train_corpus, n=1)

{('[',): 10,
 ('Letra',): 1,
 ('de',): 3,
 ('``',): 1,
 ('¡Oh',): 1,
 (',',): 11,
 ('Algoritmo',): 1,
 ('!',): 1,
 ("''",): 1,
 ('ft.',): 1,
 ('Nora',): 3,
 ('Erez',): 3,
 (']',): 10,
 ('Refrán',): 2,
 (':',): 11,
 ('Jorge',): 10,
 ('Drexler',): 10,
 ('¿Quién',): 11,
 ('quiere',): 11,
 ('que',): 38,
 ('yo',): 14,
 ('quiera',): 11,
 ('lo',): 13,
 ('creo',): 11,
 ('quiero',): 11,
 ('?',): 18,
 ('Estribillo',): 2,
 ('Dime',): 3,
 ('qué',): 3,
 ('debo',): 3,
 ('cantar',): 3,
 ('Oh',): 2,
 ('algoritmo',): 3,
 ('Sé',): 2,
 ('sabes',): 2,
 ('mejor',): 2,
 ('Incluso',): 2,
 ('mismo',): 2,
 ('Verso',): 4,
 ('1',): 1,
 ('Wait',): 1,
 ('what',): 3,
 ("'s",): 3,
 ('that',): 4,
 ('money',): 1,
 ('you',): 6,
 ('spent',): 1,
 ('What',): 1,
 ('sitting',): 1,
 ('on',): 3,
 ('your',): 1,
 ('plate',): 1,
 ('Do',): 1,
 ('want',): 4,
 ("'ve",): 1,
 ('been',): 1,
 ('fed',): 1,
 ('Are',): 1,
 ('the',): 8,
 ('fish',): 1,
 ('or',): 1,
 ('bait',): 1,
 ('Mmm',): 1,
 ('I',): 7,
 ("'m",): 2,
 ('top',): 1,
 ('of',)

In [36]:
n_grams(train_corpus, n=2)

{('BOS', '['): 10,
 ('[', 'Letra'): 1,
 ('Letra', 'de'): 1,
 ('de', '``'): 1,
 ('``', '¡Oh'): 1,
 ('¡Oh', ','): 1,
 (',', 'Algoritmo'): 1,
 ('Algoritmo', '!'): 1,
 ('!', "''"): 1,
 ("''", 'ft.'): 1,
 ('ft.', 'Nora'): 1,
 ('Nora', 'Erez'): 3,
 ('Erez', ']'): 3,
 (']', 'EOS'): 10,
 ('[', 'Refrán'): 2,
 ('Refrán', ':'): 2,
 (':', 'Jorge'): 7,
 ('Jorge', 'Drexler'): 10,
 ('Drexler', ']'): 7,
 ('BOS', '¿Quién'): 11,
 ('¿Quién', 'quiere'): 11,
 ('quiere', 'que'): 11,
 ('que', 'yo'): 13,
 ('yo', 'quiera'): 11,
 ('quiera', 'lo'): 11,
 ('lo', 'que'): 11,
 ('que', 'creo'): 11,
 ('creo', 'que'): 11,
 ('que', 'quiero'): 11,
 ('quiero', '?'): 11,
 ('?', 'EOS'): 18,
 ('[', 'Estribillo'): 2,
 ('Estribillo', ':'): 2,
 ('BOS', 'Dime'): 2,
 ('Dime', 'qué'): 3,
 ('qué', 'debo'): 3,
 ('debo', 'cantar'): 3,
 ('cantar', 'EOS'): 2,
 ('BOS', 'Oh'): 2,
 ('Oh', ','): 2,
 (',', 'algoritmo'): 2,
 ('algoritmo', 'EOS'): 2,
 ('BOS', 'Sé'): 2,
 ('Sé', 'que'): 2,
 ('que', 'lo'): 2,
 ('lo', 'sabes'): 2,
 ('sabes', 'mej

In [37]:
n_grams(train_corpus, n=3)

{('BOS', '[', 'Letra'): 1,
 ('[', 'Letra', 'de'): 1,
 ('Letra', 'de', '``'): 1,
 ('de', '``', '¡Oh'): 1,
 ('``', '¡Oh', ','): 1,
 ('¡Oh', ',', 'Algoritmo'): 1,
 (',', 'Algoritmo', '!'): 1,
 ('Algoritmo', '!', "''"): 1,
 ('!', "''", 'ft.'): 1,
 ("''", 'ft.', 'Nora'): 1,
 ('ft.', 'Nora', 'Erez'): 1,
 ('Nora', 'Erez', ']'): 3,
 ('Erez', ']', 'EOS'): 3,
 ('BOS', '[', 'Refrán'): 2,
 ('[', 'Refrán', ':'): 2,
 ('Refrán', ':', 'Jorge'): 2,
 (':', 'Jorge', 'Drexler'): 7,
 ('Jorge', 'Drexler', ']'): 7,
 ('Drexler', ']', 'EOS'): 7,
 ('BOS', '¿Quién', 'quiere'): 11,
 ('¿Quién', 'quiere', 'que'): 11,
 ('quiere', 'que', 'yo'): 11,
 ('que', 'yo', 'quiera'): 11,
 ('yo', 'quiera', 'lo'): 11,
 ('quiera', 'lo', 'que'): 11,
 ('lo', 'que', 'creo'): 11,
 ('que', 'creo', 'que'): 11,
 ('creo', 'que', 'quiero'): 11,
 ('que', 'quiero', '?'): 11,
 ('quiero', '?', 'EOS'): 11,
 ('BOS', '[', 'Estribillo'): 2,
 ('[', 'Estribillo', ':'): 2,
 ('Estribillo', ':', 'Jorge'): 2,
 ('BOS', 'Dime', 'qué'): 2,
 ('Dime', 'qué'

Debe mostrar que su método funciona para $N = 1,2,3$.

## P7. Perplexity (1 punto)

En esta sección evaluarán su modelo de n-gramas y determinarán la probabilidad de oraciones y la perplejidad con un conjunto de test. Recuerde que la perplejidad se define de la siguiente manera:

$$
\text{Perplexity} = 2^{-l} \quad \quad l = \frac{1}{M} \sum_{i=1}^{m} \log p(s_i)
$$

con $m$ el número de oraciones del corpus y $M$ el tamaño del vocabulario.

### 7.a Obtener probabilidades (0.5 puntos)

En esta sección implementará una función que determine la probabilidad de una oración.

Defina una función que reciba una oración, un diccionario con n-gramas y el valor de $n$. La función debe entregar la probabilidad de cualquier oración.

**Hint**: No olvide los posibles casos borde, como palabras fuera del vocabulario.

In [38]:
def get_probability(sentence, ngram_model, n):
    if len(sentence) < n:
        return 1e-8

    min_prob = 1e-8
    probs = []

    for i in range(len(sentence) - n + 1):
        ngram = tuple(sentence[i:i + n])
        prefix = ngram[:-1]

        count_ngram = ngram_model.get(ngram, 0)
        count_prefix = sum(v for k, v in ngram_model.items() if k[:-1] == prefix)

        p = count_ngram / count_prefix if count_prefix > 0 else min_prob
        probs.append(max(p, min_prob))

    return float(np.prod(probs))

Pruebe su función con oraciones frecuentes y comente sus resultados



In [39]:
ngram_model = n_grams(train_corpus, n=3)

In [40]:
get_probability(word_tokenize("BOS ¿Quién quiere que yo quiera lo que creo que quiero ? EOS"), ngram_model, 3)

0.8461538461538461

In [41]:
get_probability(word_tokenize("Existe algo así en el texto realmente"), ngram_model, 3)

1.0000000000000001e-40

In [42]:
get_probability(word_tokenize("Jorge Drexler  "), ngram_model, 3)

1e-08

Las tres oraciones evaluadas ilustran distintos comportamientos del modelo trigram: la primera, altamente frecuente en el corpus, obtuvo una probabilidad elevada debido a que todos sus trigramas están presentes en el modelo; la segunda, inexistente en el entrenamiento, fue penalizada severamente con una probabilidad extremadamente baja por contener solo trigramas no observados; la tercera, al ser demasiado corta para formar trigramas, activa un caso borde y retorna directamente la probabilidad mínima asignada por defecto (1e-08).

### 7.b Perplexity en conjunto de test (0.5 puntos)

En esta sub-sección deberá calcular la perplejidad del corpus de test.

Defina una función que reciba un corpus de test y retorne la perplexity (ver clases del curso). Utilice la función de la sección anterior.

In [43]:
def get_perplexity(corpus, n):
    ngram_model = n_grams(train_corpus, n)

    total_log_probs = []
    total_tokens = 0

    for sentence in corpus:
        tokens = word_tokenize(f'BOS {sentence.strip()} EOS')
        total_tokens += len(tokens)
        prob = get_probability(tokens, ngram_model, n)
        log_prob = np.log2(prob) if prob > 0 else np.log2(1e-8)
        total_log_probs.append(log_prob)

    entropy = -np.sum(total_log_probs) / total_tokens
    return float(2 ** entropy)

In [44]:
get_perplexity(test_corpus, 3)

13.98616718737452

Nosotros interpretamos el valor de perplexity ≈ 13,99 como una medida del grado de incertidumbre promedio que el modelo trigramas manifiesta al predecir las palabras del corpus de test. Este valor indica que, al enfrentarse a las secuencias lingüísticas del conjunto de prueba, el modelo considera en promedio unas 14 opciones posibles como continuaciones plausibles. La magnitud del resultado sugiere que el modelo logra capturar una porción significativa de las regularidades del lenguaje en el corpus, aunque aún conserva un margen de dispersión en sus predicciones

## P8. Interpolación Lineal (0.5 puntos)

Cree una función que obtenga la probabilidad de una oración interpolando linealmente modelos de unigrama, bigrama y trigrama ponderados por $\lambda_1, \lambda_2$ y $\lambda_3$ respectivamente. Para esto use las funciones que creó anteriormente.

In [45]:
def get_probability_lineal_interpol(sentence, corpus, l_1, l_2, l_3):
    from nltk.tokenize import word_tokenize
    from collections import Counter
    import numpy as np

    assert abs(l_1 + l_2 + l_3 - 1.0) < 1e-6, "Los pesos deben sumar 1"

    # Modelos de n-gramas
    model_1 = Counter(n_grams(corpus, 1))
    model_2 = Counter(n_grams(corpus, 2))
    model_3 = Counter(n_grams(corpus, 3))

    total_1 = sum(model_1.values())
    tokens = word_tokenize(f'BOS {sentence.strip()} EOS')
    min_prob = 1e-8
    probs = []

    for i in range(len(tokens)):
        w1 = tokens[i]
        w2 = tokens[i - 1] if i >= 1 else None
        w3 = tokens[i - 2] if i >= 2 else None

        # Unigrama
        p1 = model_1.get((w1,), 0) / total_1 if (w1,) in model_1 else 0

        # Bigrama
        if w2:
            prefix_count = sum(v for k, v in model_2.items() if k[0] == w2)
            p2 = model_2.get((w2, w1), 0) / prefix_count if prefix_count > 0 else 0
        else:
            p2 = 0

        # Trigrama
        if w2 and w3:
            trigram_prefix = (w3, w2)
            prefix_count = sum(v for k, v in model_3.items() if k[:2] == trigram_prefix)
            p3 = model_3.get((w3, w2, w1), 0) / prefix_count if prefix_count > 0 else 0
        else:
            p3 = 0

        interpolated = l_1 * p1 + l_2 * p2 + l_3 * p3
        probs.append(max(interpolated, min_prob))

    return float(np.prod(probs))


Defina una función para calcular la perplejidad de un corpus con interpolación lineal.

In [46]:
def get_pp_interpol(corpus, l_1, l_2, l_3):
    """
    Calcula la perplejidad de un corpus de prueba utilizando interpolación lineal.

    Args:
        corpus (list[str]): Lista de oraciones (test_corpus).
        l_1, l_2, l_3 (float): Pesos para unigramas, bigramas y trigramas (deben sumar 1).

    Returns:
        float: Valor de perplexity del corpus con interpolación.
    """
    import numpy as np
    from nltk.tokenize import word_tokenize

    total_log_prob = 0.0
    total_tokens = 0

    for sentence in corpus:
        tokens = word_tokenize(f'BOS {sentence.strip()} EOS')
        total_tokens += len(tokens)
        prob = get_probability_lineal_interpol(sentence, train_corpus, l_1, l_2, l_3)
        log_prob = np.log2(prob) if prob > 0 else np.log2(1e-8)
        total_log_prob += log_prob

    entropy = - total_log_prob / total_tokens
    return float(2 ** entropy)


Ahora haga pruebas con distintos valores de $\lambda_1, \lambda_2$ y $\lambda_3$, incluyendo valores extremos (por ejemplo $[1, 0, 0]$). Comente sus resultados.

## Evaluación valores arbitrarios

In [47]:
get_pp_interpol(test_corpus, 0.33, 0.33, 0.34)

22.76846587140421

In [48]:
get_pp_interpol(test_corpus, 0.1, 0.3, 0.6)

20.28072362824061

In [49]:
get_pp_interpol(test_corpus, 0.6, 0.3, 0.1)

32.268745617881706

In [50]:
get_pp_interpol(test_corpus, 0.1, 0.6, 0.3)

19.689912632063717

In [51]:
get_pp_interpol(test_corpus, 0.3, 0.6, 0.1)

22.734592387217653

Realizando estas pruebas es posible apreciar que hay combinaciones de $\lambda$ que entregan mejores resultados de perplexity que otros. Además de indicar cual es n-grama que tiene mayor peso en la interpolación.

## Evaluación valores extemos

In [72]:
get_pp_interpol(test_corpus, 1, 0, 0)

782.2597481383283

In [73]:
get_pp_interpol(test_corpus, 0, 1, 0)

39.35619117450112

In [74]:
get_pp_interpol(test_corpus, 0, 0, 1)

450.352023599737

Se puede observar que al asignar el 100% del peso a un solo modelo n-grama, la perplexity varía drásticamente según el orden considerado. El modelo de unigramas $(1, 0, 0)$ alcanza una perplexity muy alta ($782.26$), lo que refleja su incapacidad para capturar dependencias contextuales. El bigrama $(0, 1, 0)$ ofrece el mejor desempeño ($39.36$), al equilibrar contexto y generalización. En cambio, el trigrama $(0, 0, 1)$ sufre de sobreajuste, con una perplexity elevada ($450.35$), evidenciando que ningún modelo extremo es óptimo por sí solo, y que la interpolación es clave para un balance eficaz

## Extra

Por cuestiones de curiosidad quise saber el valor de todos los posibles valores sin entrar en los decimales tan exactos (aun) así que diseñe una función que encuentra la mejor interpolación dentro de todos los valores posibles de lambda:

In [55]:
def find_best_interpolation(test_corpus):
    best_perplexity = float('inf')
    best_weights = None

    for l1 in np.arange(0.0, 1.01, 0.1):
        for l2 in np.arange(0.0, 1.01 - l1, 0.1):
            l3 = round(1.0 - l1 - l2, 1)
            if l3 < 0.0 or l3 > 1.0:
                continue
            try:
                pp = get_pp_interpol(test_corpus, round(l1, 1), round(l2, 1), round(l3, 1))
                if pp < best_perplexity:
                    best_perplexity = pp
                    best_weights = (round(l1, 1), round(l2, 1), round(l3, 1))
            except Exception as e:
                continue

    return best_perplexity, best_weights


In [56]:
best_pp, best_lambdas = find_best_interpolation(test_corpus)
print(f"Mejor perplexity: {best_pp}")
print(f"Pesos óptimos: λ1={best_lambdas[0]}, λ2={best_lambdas[1]}, λ3={best_lambdas[2]}")


Mejor perplexity: 19.666278810965135
Pesos óptimos: λ1=0.1, λ2=0.5, λ3=0.4


Ahora con 2 decimales de exactitud:

In [ ]:
def refine_best_interpolation(test_corpus, best_weights, step=0.01, delta=0.05):
    best_perplexity = float('inf')
    best_refined_weights = None

    l1_center, l2_center, _ = best_weights

    # Definir rangos locales para l1 y l2
    l1_range = np.round(np.arange(max(0.0, l1_center - delta), min(1.0, l1_center + delta) + step, step), 2)

    for l1 in l1_range:
        l2_max = 1.0 - l1
        l2_range = np.round(np.arange(max(0.0, l2_center - delta), min(l2_max, l2_center + delta) + step, step), 2)

        for l2 in l2_range:
            l3 = round(1.0 - l1 - l2, 2)
            if l3 < 0.0 or l3 > 1.0:
                continue
            try:
                perplexity = get_pp_interpol(test_corpus, l1, l2, l3)
                if perplexity < best_perplexity:
                    best_perplexity = perplexity
                    best_refined_weights = (l1, l2, l3)
            except Exception:
                continue

    return best_perplexity, best_refined_weights


In [59]:
best_pp_1d, best_weights_1d = find_best_interpolation(test_corpus)

best_pp_refined, best_weights_refined = refine_best_interpolation(test_corpus, best_weights_1d)

print(f"Mejor perplexity refinada: {best_pp_refined}")
print(f"Pesos óptimos refinados: λ1={best_weights_refined[0]}, λ2={best_weights_refined[1]}, λ3={best_weights_refined[2]}")


Mejor perplexity refinada: 19.617784454120947
Pesos óptimos refinados: λ1=0.08, λ2=0.55, λ3=0.37


Curiosamente funcionó, así que para este caso, los mejores lambda son λ1=0.08, λ2=0.55, λ3=0.37 para cada modelo de n-grama respectivamente. 